designs/closures-design.md

```diff
@@ -0,0 +1,143 @@
+# Closures Design Document
+
+## New expression
+
+### Grammar
+
+```pl
+literal :=
+...
+| lambda [<name> [, <name>]*]? : <expr>
+```
+
+### AST
+
+```ts
+// To stay a subset of Python, lambda parameters cannot have type hints,
+// so we defer the type-checking of an anonymous function to the point of use
```

Comment:
Oof this seems super tricky to get right. I don't know if this check-at-use behavior is OK. Reporting the error for multiple conflicting uses is tough, and how will type-checking work if the lambda is an argument to an unresolved lambda param?

I wonder what other syntactic hacks are possible. Maybe a special variable-arity function like `mklambda(Callable[...], lambda x, y: ...)` or similar, just to have a place to put the annotation?

Reply:
That's a good point, we don't have a quick fix either and we'll think about it.